### PR TITLE
Update the Pattern Matching topic for Null

### LINKED ISSUE #1845: [C# Guide] Explain that `null` is a constant pattern, and won't match any type
Our samples on pattern matching are correct, and we mention that `null` doesn't match any type in an `is` or `case` pattern matching expression.  

[this topic](https://github.com/dotnet/docs/blob/master/docs/csharp/pattern-matching.md) should add a paragraph or two explaining the behavior for `null`. We can expand with the justifications:

1. The existing `is` expression always returned `false` for `null`. It is more consistent for the new extensions to have the same semantics on null.
2. Having `null` never match a type means that developers do not need the explicit null check after matching a pattern. That is a factor in favor of this design.
3. *If* non-nullable types are added to the language, this decision will make everyone happy.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #1845

Add a more detailed explanation as to why the null constant does not
match any type pattern. It can only match the null constant pattern.

### CODE REVIEW

File: docs/csharp/pattern-matching.md [hunk 1]
```diff
@@ -205,9 +205,17 @@ Finally, you can add a `null` case to ensure the argument is not `null`:
 
 [!code-csharp[NullCase](../../samples/csharp/PatternMatching/GeometricUtilities.cs#10_NullCase "Add null case")]
 
-The special case for the `null` pattern is interesting because the constant `null` does
-not have a type, but can be converted to any reference type or nullable
-type. 
+The special behavior for the `null` pattern is interesting because the constant
+`null` does not have a type, but can be converted to any reference
+type or nullable type. The value `null` does not match any type pattern,
+regardless of the compile time type of the variable. There are no null
```
Comment (rpetrusha) [inline docs/csharp/pattern-matching.md]: nit: compile-time

File: docs/csharp/pattern-matching.md [hunk 2]
```diff
@@ -205,9 +205,17 @@ Finally, you can add a `null` case to ensure the argument is not `null`:
...
+checks in any of the case blocks of the samples above. They would not be
```
Comment (rpetrusha) [inline docs/csharp/pattern-matching.md]: would not be --> are not

File: docs/csharp/pattern-matching.md [hunk 3]
```diff
@@ -205,9 +205,17 @@ Finally, you can add a `null` case to ensure the argument is not `null`:
...
+necessary: matching the type pattern guarantees a non null value.
+
+This special case makes the new `switch` based type pattern consistent
+with the `is` statement: `is` statements always returned `false` when
```
Comment (rpetrusha) [inline docs/csharp/pattern-matching.md]: returned --> return

File: docs/csharp/pattern-matching.md [hunk 4]
```diff
@@ -205,9 +205,17 @@ Finally, you can add a `null` case to ensure the argument is not `null`:
...
+the value being checked was `null`. It's also simpler: once you have
```
Comment (rpetrusha) [inline docs/csharp/pattern-matching.md]: was --> is

Comment (rpetrusha) [inline docs/csharp/pattern-matching.md]: have checked --> check

File: docs/csharp/pattern-matching.md [hunk 5]
```diff
@@ -205,9 +205,17 @@ Finally, you can add a `null` case to ensure the argument is not `null`:
 
 [!code-csharp[NullCase](../../samples/csharp/PatternMatching/GeometricUtilities.cs#10_NullCase "Add null case")]
 
-The special case for the `null` pattern is interesting because the constant `null` does
-not have a type, but can be converted to any reference type or nullable
-type. 
+The special behavior for the `null` pattern is interesting because the constant
+`null` does not have a type, but can be converted to any reference
+type or nullable type. The value `null` does not match any type pattern,
```
Comment (rpetrusha) [inline docs/csharp/pattern-matching.md]: I'm not sure why, but I find this sentence "The value `null`..." difficult to understand." and the paragraph too compact and disconnected. Perhaps: 
The special behavior for the `null` pattern is interesting because the constant `null` in the pattern does not have a type but can be converted to any reference type or nullable type. Without this pattern, a `null` value would not match any type pattern, regardless of the compile-time type of the variable. We can see that from the fact that there are no null checks in any of the case blocks of the samples above: they are not necessary, since matching the type pattern guarantees a non-null value.

**Review by rpetrusha — CHANGES_REQUESTED** [review]
@BillWagner, a number of small comments, but a suggested rewording for some text that I found hard to parse.

File: docs/csharp/pattern-matching.md (hunk 5 above)
Comment (BillWagner) [inline docs/csharp/pattern-matching.md]: I took a lot of the text you suggested, and combined this paragraph with the next one. I think it reads much better now. Thanks for the suggestion.

File: docs/csharp/pattern-matching.md (hunk 1 above)
Comment (BillWagner) [inline docs/csharp/pattern-matching.md]: Fixed.

File: docs/csharp/pattern-matching.md (hunk 2 above)
Comment (BillWagner) [inline docs/csharp/pattern-matching.md]: fixed

File: docs/csharp/pattern-matching.md (hunk 3 above)
Comment (BillWagner) [inline docs/csharp/pattern-matching.md]: fixed.

**Comment by BillWagner** [comment]
@rpetrusha Thanks for the review. I updated the paragraph, using most of your suggestions and combining it with the following paragraph. It's ready for another review.

**Review by rpetrusha — APPROVED** [review]
@BillWagner, LGTM. Let's merge.
